cpp/tests/io/parquet_chunked_reader_test.cu

```diff
@@ -953,64 +953,283 @@ TEST_F(ParquetChunkedReaderTest, TestChunkedReadNullCount)
+                             cudf::io::compression_type::ZSTD,
+                             cudf::io::dictionary_policy::NEVER);
+  // compression with a codec that uses no scratch space at decode time
+  input_limit_test_write_one(base_path + "_c.parquet",
+                             t,
+                             cudf::io::compression_type::SNAPPY,
+                             cudf::io::dictionary_policy::NEVER);
+  input_limit_test_write_one(base_path + "_d.parquet",
+                             t,
+                             cudf::io::compression_type::SNAPPY,
+                             cudf::io::dictionary_policy::ALWAYS);
+}
 
-  {
-    // 25 chunks of 100k rows each
-    auto const [result, num_chunks] = chunked_read(filepath, 0, 4000000);
-    EXPECT_EQ(num_chunks, 25);
-    CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected, result->get_column(0));
-  }
+void input_limit_test_read(std::string const& base_path,
+                           cudf::table_view const& t,
+                           size_t output_limit,
+                           size_t input_limit,
+                           int const expected_chunk_counts[4])
+{
+  std::vector<std::string> file_suffix{"_a.parquet", "_b.parquet", "_c.parquet", "_d.parquet"};
+  CUDF_EXPECTS(file_suffix.size() == 4,
+               "Unexpected mismatch between number of test cases and result count");
 
-  {
-    // 25 chunks of 100k rows each
-    auto const [result, num_chunks] = chunked_read(filepath, 0, 4100000);
-    EXPECT_EQ(num_chunks, 25);
-    CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected, result->get_column(0));
+  for (size_t idx = 0; idx < file_suffix.size(); idx++) {
+    auto result = chunked_read(base_path + file_suffix[idx], output_limit, input_limit);
+    CUDF_EXPECTS(result.second == expected_chunk_counts[idx],
+                 "Unexpected number of chunks produced in chunk read");
+    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(*result.first, t);
   }
+}
 
-  {
-    // 12 chunks of 200k rows each, plus 1 final chunk of 100k rows.
-    auto const [result, num_chunks] = chunked_read(filepath, 0, 8002301);
-    EXPECT_EQ(num_chunks, 13);
-    CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected, result->get_column(0));
-  }
+struct ParquetChunkedReaderInputLimitConstrainedTest : public cudf::test::BaseFixture {};
+
+TEST_F(ParquetChunkedReaderInputLimitConstrainedTest, SingleFixedWidthColumn)
+{
+  auto base_path          = temp_env->get_temp_filepath("single_col_fixed_width");
+  constexpr auto num_rows = 1'000'000;
+  auto iter1 = cudf::detail::make_counting_transform_iterator(0, [](int i) { return 15; });
+  cudf::test::fixed_width_column_wrapper<double> col1(iter1, iter1 + num_rows);
+  auto tbl = cudf::table_view{{col1}};
+
+  input_limit_test_write(base_path, tbl);
+
+  // semi-reasonable limit
+  constexpr int expected_a[] = {1, 17, 4, 1};
+  input_limit_test_read(base_path, tbl, 0, 2 * 1024 * 1024, expected_a);
+  // an unreasonable limit
+  constexpr int expected_b[] = {1, 50, 50, 1};
```

Comment:
how come we only get a single chunk in some of these cases?

Reply:
So in the initial case (array entry 0) the data is uncompressed - so no need for temporary space. There's only 1 row group, so the entire blob of data is ~8MB. We have no choice but to load that.  But once we do, there's no decompression to do so we can return the whole thing in 1 pass.

And the last case includes a dictionary, which obliterates the data size even before compression.  The whole row group is just a bit over 1KB.